### ISSUE TITLE
在计算attention score的时候如何对padding做mask操作?

### ISSUE THREAD

**Issue by Sogrey** [Sogrey/AI-QA#565]
在计算注意力得分（attention score）时，对padding进行mask操作是自然语言处理（NLP）中常见的技术，目的是避免模型在计算注意力权重时对填充（padding）部分进行不必要的关注。以下是一个详细的解释和操作步骤：

------

### 📌 **1. 什么是Padding Mask？**

在处理变长序列（如句子）时，通常会用padding将所有序列填充到相同的长度。这些填充部分在实际中是没有意义的，因此在计算注意力得分时，需要通过mask操作将这些位置的注意力权重设置为负无穷（-inf），这样在softmax函数中，这些位置的权重会趋近于0，从而不会对最终的注意力结果产生影响。

------

### 🧮 **2. 实现步骤**

1. **创建mask矩阵**：
   - 一个二维矩阵，形状为（batch_size, sequence_length）。
   - 矩阵中，padding位置的值为1，有效位置的值为0。

2. **应用mask到注意力得分**：
   - 将mask矩阵与注意力得分矩阵相乘，其中mask为1的位置将注意力得分设置为负无穷。
   - 在PyTorch中，可以使用`torch.where`函数实现这一操作。

3. **应用softmax函数**：
   - 对mask后的注意力得分应用softmax函数，使其在每行上归一化。

------

### 🧠 **3. 示例代码（PyTorch）**

```python
import torch

# 假设batch_size=2, sequence_length=5, num_heads=2
# 注意力得分矩阵（形状为 (batch_size, num_heads, seq_len, seq_len)）
attention_scores = torch.randn(2, 2, 5, 5)

# 创建padding mask（假设padding位置为最后两个位置）
padding_mask = torch.tensor([[0, 0, 0, 0, 1], [0, 0, 0, 0, 1]])

# 应用mask
masked_scores = torch.where(padding_mask == 1, -float('inf'), attention_scores)

# 应用softmax函数
softmax_scores = torch.nn.functional.softmax(masked_scores, dim=-1)
```

------

### 📌 **4. 总结**

通过padding mask操作，可以确保模型在计算注意力权重时忽略填充部分，从而提高模型的性能和准确性。这一操作在处理变长序列时尤为重要，是NLP中常见的技巧之一。

> 💡 **注意**：在实际应用中，padding mask的创建需要根据具体的数据格式和填充方式来调整。

-----

来自n8n
